src/target/riscv/riscv-013.c

```diff
@@ -2027,6 +2026,10 @@ static int examine(struct target *target)
 		return ERROR_FAIL;
 	}
 
+	/* The RISC-V hartid is sequential, and the index of each hart
+	 * on the Debug Module should start at 0 and be contiguous. */
+	info->index = target->coreid % dm->hart_count;
```

Comment:
Doesn't this assume that each DM has the same number of harts?

Reply:
Thanks for your reminder,  I didn't consider the situation that each DM has different number of harts, if this situation happens, user configuration will become unfriendly. So I currently have two ideas to solve it:
**Config modification**
config like
```
# cluster 0
target create $_TARGETNAME_0 riscv -chain-position $_CHIPNAME.cpu -coreid 0 -threadid 0 -rtos hwthread
target create $_TARGETNAME_1 riscv -chain-position $_CHIPNAME.cpu -coreid 1 -threadid 1

# cluster 1
target create $_TARGETNAME_2 riscv -chain-position $_CHIPNAME.cpu -coreid 0 -threadid 2 -dbgbase 0x400 -rtos hwthread
target create $_TARGETNAME_3 riscv -chain-position $_CHIPNAME.cpu -coreid 1 -threadid 3 -dbgbase 0x400

target smp $_TARGETNAME_0 $_TARGETNAME_1 $_TARGETNAME_2 $_TARGETNAME_3
```
Modify the 'jim_target_create' function to get the target->threadid.

threadid define
https://github.com/riscv/riscv-openocd/blob/21d21408aa36ccb458ccfaad871f7dd1bb1f731a/src/rtos/hwthread.c#L40-L43
changed to
```
static inline threadid_t threadid_from_target(const struct target *target)
{
	return target->threadid + 1;
}
```
**Get different threadids through hartid**
config like
```
# cluster 0
target create $_TARGETNAME_0 riscv -chain-position $_CHIPNAME.cpu -coreid 0 -rtos hwthread
target create $_TARGETNAME_1 riscv -chain-position $_CHIPNAME.cpu -coreid 1

# cluster 1
target create $_TARGETNAME_2 riscv -chain-position $_CHIPNAME.cpu -coreid 0 -dbgbase 0x400 -rtos hwthread
target create $_TARGETNAME_3 riscv -chain-position $_CHIPNAME.cpu -coreid 1 -dbgbase 0x400

target smp $_TARGETNAME_0 $_TARGETNAME_1 $_TARGETNAME_2 $_TARGETNAME_3
```
Make appropriate modifications to the "threadid_from_target' function, I still don’t know how to modify it
```
static inline threadid_t threadid_from_target(const struct target *target)
{
	#return target->coreid + 1;
        ...
}
```
thread id should be 0/1/2/3